File: chazm-model/src/main/java/runtimemodels/chazm/model/relation/RequiresFactory.java
package runtimemodels.chazm.model.relation;

import runtimemodels.chazm.api.entity.Capability;
import runtimemodels.chazm.api.entity.Role;
import runtimemodels.chazm.api.relation.Requires;

/**
 * The {@linkplain RequiresFactory} interface defines the API for constructing {@linkplain Requires} relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface RequiresFactory {

    /**
     * Constructs a {@linkplain Requires}.
     *
     * @param role       the {@linkplain Role} of the {@linkplain Requires}.
     * @param capability the {@linkplain Capability} of the {@linkplain Requires}.
     * @return a {@linkplain Requires}.
     */
    Requires buildRequires(Role role, Capability capability);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/RequiresFactory.kt
package runtimemodels.chazm.model.relation

import runtimemodels.chazm.api.entity.Capability
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.relation.Requires

/**
 * The [RequiresFactory] interface defines the API for constructing [Requires] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface RequiresFactory {
    /**
     * Constructs a [Requires].
     *
     * @param role       the [Role] of the [Requires].
     * @param capability the [Capability] of the [Requires].
     * @return a [Requires].
     */
    fun build(role: Role, capability: Capability): Requires
}
